### ISSUE TITLE
Critical performance issues

### ISSUE THREAD

**Issue by Hellenic** [neo4j/graphql#3247]
**Describe the bug**
We have quite large schema with many different relationships, but not that much data yet. In my free Aura instance test data set is about 1% of available node capacity and 2% of relationships. With this amount of data most of our queries take seconds to load. We were able to also pinpoint a query which totally blocks Neo4j smaller instances (and crashes the free instance). I would be happy to show this to you.

**Type definitions**
Here's a simple gist to somewhat reproduce the issues; https://gist.github.com/Hellenic/a70657ca3f4abfbaad1c24e968277cb9
It generates rougly 200 nodes and 400 relationships. When running that against Neo4j Aura free, the query takes 2+ seconds, even with that small amount of data. Commenting out the @auth directives, it runs in rougle 500ms, which is still quite slow but huge different to the above. Scaling this simple example to actual system, the query times become unbearable.

**Expected behavior**
I would expect the queries be a lot faster, as that's exactly what I understand Neo4j is supposed to do well; Nodes and relationships. It would be good to understand are we doing something wrong, and how this can be improved.

**System (please complete the following information):**

* OS: Linux, MacOS
* Version:@neo4j/graphql@3.16.1, neo4j-driver@5.5.0, @neo4j/graphql-plugin-auth@2.1.0
* Node.js version: 16.19.1

**Comment by neo4j-team-graphql** [comment]
Many thanks for raising this bug report @Hellenic. :bug: We will now attempt to reproduce the bug based on the steps you have provided.

Please ensure that you've provided the necessary information for a minimal reproduction, including but not limited to:
* Type definitions
* Resolvers
* Query and/or Mutation (or multiple) needed to reproduce

If you have a support agreement with Neo4j, please link this GitHub issue to a new or existing Zendesk ticket.

Thanks again! :pray:

**Comment by neo4j-team-graphql** [comment]
Many thanks for raising this bug report @Hellenic. :bug: We will now attempt to reproduce the bug based on the steps you have provided.

Please ensure that you've provided the necessary information for a minimal reproduction, including but not limited to:
* Type definitions
* Resolvers
* Query and/or Mutation (or multiple) needed to reproduce

If you have a support agreement with Neo4j, please link this GitHub issue to a new or existing Zendesk ticket.

Thanks again! :pray:

**Comment by darrellwarde** [comment]
Hey @Hellenic! A bit of background here:

The current `allow` and `bind` rules of `@auth` are evaluated using a mixture of `apoc.util.validate` and `apoc.util.validatePredicate`. Furthermore, the predicates are currently generated completely separately to the predicates which are generated when using the `where` argument in a query.

As such, the predicate generation has not gone through some of the optimizations that we've made to the Cypher generated for general filters. My _suspicion_ here is that the poorly performing queries are when checking auth rules that refer to different nodes over a relationship.

We are in the process of completely rearchitecting the authorization functionality, and the generation of filtering predicates is now aligned across all features. This is still a work in progress, but thanks to your excellent reproduction instructions, we'll be able to create a performance test to check the difference between the current and rearchitected implementations of this functionality to see if we inadvertently fix the issue.

If we still witness problems, we'll go back to the drawing board.

**Comment by Hellenic** [comment]
Hello @darrellwarde and thank you for quick response!
Thanks for the explanation. It's true `where` might actually work better for us, and great if it helps with the performance. I will try some changes in the upcoming days and report back how it goes.

Looking forward to the next major version too with the authorization changes. Do you have any timeline on that?

**Comment by darrellwarde** [comment]
We have a rough timeline of the middle of the year - the team is working hard to get the next major version out by the end of June, but this milestone has flexibility to make sure we get it right!

**Comment by Hellenic** [comment]
Hey! Thank you. We attempted an upgrade but were unable to continue. We came up with a new bug, introduced in some of the versions (we upgraded from `3.16.1`). When we make a mutation and if we only request the `info` object, the request fails with: 
```
Invalid input ',': expected \"*\", \"DISTINCT\" or an expression (line 62, column 8 (offset: 2679))\n\"RETURN , collect(DISTINCT m) as meta\"\n        ^"
```

Response body example:
```
{
    info {
      nodesCreated
      nodesDeleted
    }
}
```
We can include for example changed IDs on return bodies, and we might change to do that. But until then, cannot report on the performance just yet.

**Comment by darrellwarde** [comment]
> Hey! Thank you. We attempted an upgrade but were unable to continue. We came up with a new bug, introduced in some of the versions (we upgraded from `3.16.1`). When we make a mutation and if we only request the `info` object, the request fails with:
> 
> ```
> Invalid input ',': expected \"*\", \"DISTINCT\" or an expression (line 62, column 8 (offset: 2679))\n\"RETURN , collect(DISTINCT m) as meta\"\n        ^"
> ```
> 
> Response body example:
> 
> ```
> {
>     info {
>       nodesCreated
>       nodesDeleted
>     }
> }
> ```
> 
> We can include for example changed IDs on return bodies, and we might change to do that. But until then, cannot report on the performance just yet.

Not good! We actually have an entire integration test file for these cases of only returning the info: https://github.com/neo4j/graphql/blob/dev/packages/graphql/tests/integration/issues/only-info.int.test.ts

So could you please provide some steps to reproduce this given that it's obviously something different to what we have in our tests? Thanks! Perhaps in a new bug report if you wouldn't mind. 🙂

**Comment by darrellwarde** [comment]
> Hey! Thank you. We attempted an upgrade but were unable to continue. We came up with a new bug, introduced in some of the versions (we upgraded from `3.16.1`). When we make a mutation and if we only request the `info` object, the request fails with:
> 
> ```
> Invalid input ',': expected \"*\", \"DISTINCT\" or an expression (line 62, column 8 (offset: 2679))\n\"RETURN , collect(DISTINCT m) as meta\"\n        ^"
> ```
> 
> Response body example:
> 
> ```
> {
>     info {
>       nodesCreated
>       nodesDeleted
>     }
> }
> ```
> 
> We can include for example changed IDs on return bodies, and we might change to do that. But until then, cannot report on the performance just yet.

Hey @Hellenic, we've now fixed the above bug in 3.20.0 - hopefully this unblocks you from upgrading and providing us with some feedback! 🙂

**Comment by Hellenic** [comment]
Hey, a quick update. We now have that latest version up and running, and while there might be some queries which are now different, we don't see any huge changes. And there are still queries that can make the database run out of heap.

We've managed to get around of the most critical issues by refactoring our own quries, splitting them, and with some help from this latest version. While the situation is not yet on very solid path, it's not going to much forward with this ticket I think.

Thanks for all the help this far. I'll close this ticket, and if we can point out some specific issues in the future we'll open up a new one on those. Looking forward to the next major version!
